Config set port panic out

[kind/bug]

<!--

Welcome! - We kindly ask you to:

  1. Fill out the issue template below 
  2. Use the chat and talk to us if you have a question rather than a bug or feature request.

The chat room is at: https://chat.openshift.io/developers/channels/odo

Thanks for understanding, and for contributing to the project!

-->

## What versions of software are you using?
- Operating System: 
- Output of `odo version`: master


## How did you run odo exactly?
$ odo config set Ports 8080/TCP  - panic out
$ odo config set ports 8080/TCP - panic out
$ odo config set --ports 8080/TCP - unknown flag: --ports

## Actual behavior
Fails

## Expected behavior
Should set

## Any logs, error output, etc?
```
$ odo config set ports 8080/TCP
panic: interface conversion: interface {} is string, not []string

goroutine 1 [running]:
github.com/openshift/odo/pkg/config.(*LocalConfigInfo).SetConfiguration(0xc00013efd0, 0x7ffeefbffbcf, 0x5, 0x2296560, 0xc000af3410, 0x0, 0x2593d53)
        /Users/amit/go/src/github.com/openshift/odo/pkg/config/config.go:184 +0x103b
github.com/openshift/odo/pkg/odo/cli/config.(*SetOptions).Run(0xc0004ebe00, 0x0, 0x0)
        /Users/amit/go/src/github.com/openshift/odo/pkg/odo/cli/config/set.go:106 +0x105
github.com/openshift/odo/pkg/odo/genericclioptions.GenericRun(0x28182a0, 0xc0004ebe00, 0xc0007acb40, 0xc0006c0180, 0x2, 0x2)
        /Users/amit/go/src/github.com/openshift/odo/pkg/odo/genericclioptions/runnable.go:17 +0xf2
github.com/openshift/odo/pkg/odo/cli/config.NewCmdSet.func2(0xc0007acb40, 0xc0006c0180, 0x2, 0x2)
        /Users/amit/go/src/github.com/openshift/odo/pkg/odo/cli/config/set.go:142 +0x5e
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).execute(0xc0007acb40, 0xc0006c0120, 0x2, 0x2, 0xc0007acb40, 0xc0006c0120)
        /Users/amit/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:702 +0x289
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc00059b440, 0x0, 0x0, 0x4)
        /Users/amit/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:783 +0x29f
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).Execute(...)
        /Users/amit/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:736
main.main()
        /Users/amit/go/src/github.com/openshift/odo/cmd/odo/odo.go:73 +0x489
```
```
$ odo config set --ports 8080/TCP
Error: unknown flag: --ports
Usage:
  odo config set [flags]

Examples:

  # Set a configuration value in the local config
  odo config set Type java
  odo config set Name test
  odo config set MinMemory 50M
  odo config set MaxMemory 500M
  odo config set Memory 250M
  odo config set Ignore false
  odo config set MinCPU 0.5
  odo config set MaxCPU 2
  odo config set CPU 1
  
  # Set a env variable in the local config
  odo config set Type --env PORT=4000
  odo config set Name --env DB_USERNAME=postgres,DB_HOSTNAME=postgres
  odo config set MinMemory --env KAFKA_HOST=kafka --env KAFKA_PORT=6639

Flags:
  -e, --env stringSlice   Set the environment variables in config
  -f, --force             Don't ask for confirmation, set the config directly
  -h, --help              help for set

Global Flags:
      --alsologtostderr                  log to standard error as well as files
      --log_backtrace_at traceLocation   when logging hits line file:N, emit a stack trace (default :0)
      --log_dir string                   If non-empty, write log files in this directory
      --logtostderr                      log to standard error instead of files (default false)
      --skip-connection-check            Skip cluster check
      --stderrthreshold severity         logs at or above this threshold go to stderr (default 2)
  -v, --v Level                          log level for V logs. Level varies from 0 to 9 (default 0).
      --vmodule moduleSpec               comma-separated list of pattern=N settings for file-filtered logging

 ✗  unknown flag: --ports
```

--- a/pkg/config/config.go
+++ b/pkg/config/config.go
@@ -186,7 +186,7 @@ func (lci *LocalConfigInfo) SetConfiguration(parameter string, value interface{}
 		case "sourcelocation":
 			lci.componentSettings.SourceLocation = &strValue
 		case "ports":
-			arrValue := value.([]string)
+			arrValue := strings.Split(strValue, ",")
 			lci.componentSettings.Ports = &arrValue
 		case "name":
 			lci.componentSettings.Name = &strValue

--- a/pkg/config/config_test.go
+++ b/pkg/config/config_test.go
@@ -21,8 +21,15 @@ func TestSetLocalConfiguration(t *testing.T) {
 	minCPUValue := "0.5"
 	maxCPUValue := "2"
 	minMemValue := "500M"
+	maxMemValue := "1000M"
 	testValue := "test"
-	ports := "8080/TCP,45/UDP"
+	portsValue := "8080/TCP,45/UDP"
+	typeValue := "nodejs"
+	applicationValue := "odotestapp"
+	projectValue := "odotestproject"
+	sourceTypeValue := "git"
+	sourceLocationValue := "https://github.com/sclorg/nodejs-ex"
+	refValue := "develop"
 
 	tests := []struct {
 		name           string
@@ -80,9 +87,65 @@ func TestSetLocalConfiguration(t *testing.T) {
 			},
 		},
 		{
-			name:      fmt.Sprintf("Case 7: %s set to %s", Ports, ports),
+			name:      fmt.Sprintf("Case 7: %s set to %s", MaxMemory, maxCPUValue),
+			parameter: MaxMemory,
+			value:     maxMemValue,
+			existingConfig: LocalConfig{
+				componentSettings: ComponentSettings{},
+			},
+		},
+		{
+			name:      fmt.Sprintf("Case 8: %s set to %s", Ports, portsValue),
 			parameter: Ports,
-			value:     ports,
+			value:     portsValue,
+			existingConfig: LocalConfig{
+				componentSettings: ComponentSettings{},
+			},
+		},
+		{
+			name:      fmt.Sprintf("Case 9: %s set to %s", Type, typeValue),
+			parameter: Type,
+			value:     typeValue,
+			existingConfig: LocalConfig{
+				componentSettings: ComponentSettings{},
+			},
+		},
+		{
+			name:      fmt.Sprintf("Case 10: %s set to %s", Application, applicationValue),
+			parameter: Application,
+			value:     applicationValue,
+			existingConfig: LocalConfig{
+				componentSettings: ComponentSettings{},
+			},
+		},
+		{
+			name:      fmt.Sprintf("Case 11: %s set to %s", Project, projectValue),
+			parameter: Project,
+			value:     projectValue,
+			existingConfig: LocalConfig{
+				componentSettings: ComponentSettings{},
+			},
+		},
+		{
+			name:      fmt.Sprintf("Case 12: %s set to %s", SourceType, sourceTypeValue),
+			parameter: SourceType,
+			value:     sourceTypeValue,
+			existingConfig: LocalConfig{
+				componentSettings: ComponentSettings{},
+			},
+		},
+		{
+			name:      fmt.Sprintf("Case 12: %s set to %s", SourceLocation, sourceLocationValue),
+			parameter: SourceLocation,
+			value:     sourceLocationValue,
+			existingConfig: LocalConfig{
+				componentSettings: ComponentSettings{},
+			},
+		},
+		{
+			name:      fmt.Sprintf("Case 13: %s set to %s", Ref, refValue),
+			parameter: Ref,
+			value:     refValue,
 			existingConfig: LocalConfig{
 				componentSettings: ComponentSettings{},
 			},

--- a/pkg/config/config_test.go
+++ b/pkg/config/config_test.go
@@ -22,6 +22,7 @@ func TestSetLocalConfiguration(t *testing.T) {
 	maxCPUValue := "2"
 	minMemValue := "500M"
 	testValue := "test"
+	ports := "8080/TCP,45/UDP"
 
 	tests := []struct {
 		name           string
@@ -74,6 +75,14 @@ func TestSetLocalConfiguration(t *testing.T) {
 			name:      fmt.Sprintf("Case 6: %s set to %s", MinMemory, minMemValue),
 			parameter: MinMemory,
 			value:     minMemValue,
+			existingConfig: LocalConfig{
+				componentSettings: ComponentSettings{},
+			},
+		},
+		{
+			name:      fmt.Sprintf("Case 7: %s set to %s", Ports, ports),
+			parameter: Ports,
+			value:     ports,
 			existingConfig: LocalConfig{
 				componentSettings: ComponentSettings{},
 			},

--- a/pkg/odo/cli/config/view.go
+++ b/pkg/odo/cli/config/view.go
@@ -4,6 +4,7 @@ import (
 	"fmt"
 	"os"
 	"reflect"
+	"strings"
 	"text/tabwriter"
 
 	"github.com/openshift/odo/pkg/config"
@@ -49,6 +50,12 @@ func (o *ViewOptions) Run() (err error) {
 	w := tabwriter.NewWriter(os.Stdout, 5, 2, 2, ' ', tabwriter.TabIndent)
 	fmt.Fprintln(w, "PARAMETER", "\t", "CURRENT_VALUE")
 	fmt.Fprintln(w, "Type", "\t", showBlankIfNil(cs.Type))
+	fmt.Fprintln(w, "Application", "\t", showBlankIfNil(cs.Application))
+	fmt.Fprintln(w, "Project", "\t", showBlankIfNil(cs.Project))
+	fmt.Fprintln(w, "SourceType", "\t", showBlankIfNil(cs.SourceType))
+	fmt.Fprintln(w, "Ref", "\t", showBlankIfNil(cs.Ref))
+	fmt.Fprintln(w, "SourceLocation", "\t", showBlankIfNil(cs.SourceLocation))
+	fmt.Fprintln(w, "Ports", "\t", formatArray(cs.Ports))
 	fmt.Fprintln(w, "Name", "\t", showBlankIfNil(cs.Name))
 	fmt.Fprintln(w, "MinMemory", "\t", showBlankIfNil(cs.MinMemory))
 	fmt.Fprintln(w, "MaxMemory", "\t", showBlankIfNil(cs.MaxMemory))
@@ -73,6 +80,15 @@ func showBlankIfNil(intf interface{}) interface{} {
 	}
 
 	return intf
+}
+func formatArray(arr *[]string) string {
+	if arr == nil {
+		return ""
+	}
+	if len(*arr) == 0 {
+		return ""
+	}
+	return strings.Join(*arr, ",")
 }
 
 // NewCmdView implements the config view odo command

--- a/tests/e2e/e2e_test.go
+++ b/tests/e2e/e2e_test.go
@@ -344,14 +344,15 @@ var _ = Describe("odoe2e", func() {
 		})
 	})
 
-	Context("creating component without an application", func() {
+	Context("creating component without an application and url", func() {
 		It("should create the component in default application", func() {
 			runCmdShouldPass("odo login --username developer --password developer")
 			runCmdShouldPass("odo create php testcmp --app e2e-xyzk --git " + testPHPGitURL)
+			runCmdShouldPass("odo config set Ports 8080/TCP")
 			runCmdShouldPass("odo push")
-
+			stdOut := runCmdShouldPass("odo url create")
+			Expect(stdOut).To(ContainSubstring("8080"))
 			VerifyCmpName("testcmp")
-
 			VerifyAppNameOfComponent("testcmp", "e2e-xyzk")
 		})
 		// Uncommment after fixing the component delete once it has been modified to work with

--- a/tests/e2e/e2e_test.go
+++ b/tests/e2e/e2e_test.go
@@ -267,6 +267,8 @@ var _ = Describe("odoe2e", func() {
 				runCmdShouldPass(fmt.Sprintf("odo config set %s %s -f", testCase.paramName, testCase.paramValue))
 				Value := getConfigValue(testCase.paramName)
 				Expect(Value).To(ContainSubstring(testCase.paramValue))
+				// cleanup
+				runCmdShouldPass(fmt.Sprintf("odo config unset %s -f", testCase.paramName))
 			}
 		})
 
@@ -288,8 +290,40 @@ var _ = Describe("odoe2e", func() {
 					paramValue: "0.2",
 				},
 				{
+					paramName:  "MaxCPU",
+					paramValue: "2",
+				},
+				{
 					paramName:  "MinMemory",
 					paramValue: "100M",
+				},
+				{
+					paramName:  "MaxMemory",
+					paramValue: "500M",
+				},
+				{
+					paramName:  "Ports",
+					paramValue: "8080/TCP,45/UDP",
+				},
+				{
+					paramName:  "Application",
+					paramValue: "odotestapp",
+				},
+				{
+					paramName:  "Project",
+					paramValue: "odotestproject",
+				},
+				{
+					paramName:  "SourceType",
+					paramValue: "git",
+				},
+				{
+					paramName:  "Ref",
+					paramValue: "develop",
+				},
+				{
+					paramName:  "SourceLocation",
+					paramValue: "https://github.com/sclorg/nodejs-ex",
 				},
 			}
 

--- a/tests/e2e/e2e_test.go
+++ b/tests/e2e/e2e_test.go
@@ -227,8 +227,40 @@ var _ = Describe("odoe2e", func() {
 					paramValue: "0.2",
 				},
 				{
+					paramName:  "MaxCPU",
+					paramValue: "2",
+				},
+				{
 					paramName:  "MinMemory",
 					paramValue: "100M",
+				},
+				{
+					paramName:  "MaxMemory",
+					paramValue: "500M",
+				},
+				{
+					paramName:  "Ports",
+					paramValue: "8080/TCP,45/UDP",
+				},
+				{
+					paramName:  "Application",
+					paramValue: "odotestapp",
+				},
+				{
+					paramName:  "Project",
+					paramValue: "odotestproject",
+				},
+				{
+					paramName:  "SourceType",
+					paramValue: "git",
+				},
+				{
+					paramName:  "Ref",
+					paramValue: "develop",
+				},
+				{
+					paramName:  "SourceLocation",
+					paramValue: "https://github.com/sclorg/nodejs-ex",
 				},
 			}
 			for _, testCase := range cases {

